lib/cockpit-components-tooltip.jsx

```diff
@@ -0,0 +1,147 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2016 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+"use strict";
+
+var React = require('react');
+
+/* A tooltip, styled via Patternfly/Bootstrap.
+ *
+ * Example:
+ *
+ *   <Tooltip tip="Insufficient privileges">
+ *      <button disabled>Reboot</button>
+ *   </ToolTip>
+ *
+ * Whenever the mouse hovers over the children of a Tooltip,
+ * the text (or arbitrary element) in the "tip" property is shown.
+ */
+
+var Tooltip = React.createClass({
+    getInitialState: function () {
+        return { open: false };
+    },
+    onMouseover: function () {
+        this.setState({ open: true });
+    },
+    onMouseout: function () {
+        this.setState({ open: false });
+    },
+    render: function () {
+        var self = this;
+
+        if (!self.props.tip)
+            return self.props.children;
+
+        // Placement
+        //
+        // We assume that there is enough space to display the tooltip
+        // in its natural width, and that there is either enough space
+        // above or below the element that will show the tooltip.
+        //
+        // If there is enough space above, the tooltip goes to the
+        // top, otherwise below.  It is then moved left or right until
+        // it is fully visible.
+
+        // The life of a tooltip
+        //
+        // The tooltip starts out at position (-10000,0) so that it is
+        // well outside its parent.  This seems to give it its natural
+        // width without the browser trying to fit it into the parent,
+        // which usually results in a undesired tall and narrow
+        // layout.
+        //
+        // On first interaction and before moving it into place, we
+        // prevent the tooltip from resizing by setting the "width"
+        // and "height" styles to its current dimensions.
+        //
+        // Once moved into place, it stays where it is and is just
+        // raised and lowered in the Z dimension in addition to being
+        // made visible and invisible.
+
+        function fixup(tip) {
```

Comment:
How about calling this something more descriptive, e.g. `fixPosition`.

Is it worth explicitly returning quickly in case this is called when the component unmounts?
```
// do nothing when unmounting
if (tip === null)
    return;
```